File: packages/shared/src/gcp/secret-manager.ts
/**
 * Secret Manager Client - TypeScript wrapper for GCP Secret Manager
 *
 * Provides secure credential retrieval with local development fallback.
 */
import { SecretManagerServiceClient } from "@google-cloud/secret-manager";

const PROJECT_ID = process.env.GCP_PROJECT_ID || "viralcue-local";
const USE_SECRET_MANAGER = process.env.USE_SECRET_MANAGER === "true";

let client: SecretManagerServiceClient | null = null;

function getClient(): SecretManagerServiceClient {
  if (!client) {
    client = new SecretManagerServiceClient();
  }
  return client;
}

/**
 * Get a secret value from Secret Manager or environment variable.
 *
 * @param secretId - The secret ID in Secret Manager
 * @param envFallback - Environment variable name for local fallback
 * @param version - Version to retrieve (default: "latest")
 * @returns The secret value
 */
export async function getSecret(
  secretId: string,
  envFallback?: string,
  version: string = "latest"
): Promise<string> {
  // Local development fallback
  if (!USE_SECRET_MANAGER) {
    const envVar = envFallback || secretId.toUpperCase().replace(/-/g, "_");
    const value = process.env[envVar] || "";
    if (!value) {
      console.warn(`[SecretManager] Warning: ${envVar} not set in environment`);
    }
    return value;
  }

  const name = `projects/${PROJECT_ID}/secrets/${secretId}/versions/${version}`;

  try {
    const [response] = await getClient().accessSecretVersion({ name });
    return response.payload?.data?.toString() || "";
  } catch (error) {
    console.error(`[SecretManager] Error accessing ${secretId}:`, error);
    // Fallback to environment variable
    const envVar = envFallback || secretId.toUpperCase().replace(/-/g, "_");
    return process.env[envVar] || "";
  }
}

/**
 * Create a new secret (for initialization/migration).
 */
export async function createSecret(
  secretId: string,
  secretValue: string
): Promise<boolean> {
  if (!USE_SECRET_MANAGER) {
    console.log(`[SecretManager] Skipping create in local mode: ${secretId}`);
    return true;
  }

  const parent = `projects/${PROJECT_ID}`;

  try {
    // Create the secret
    await getClient().createSecret({
      parent,
      secretId,
      secret: { replication: { automatic: {} } },
    });

    // Add the secret version
    const secretName = `${parent}/secrets/${secretId}`;
    await getClient().addSecretVersion({
      parent: secretName,
      payload: { data: Buffer.from(secretValue, "utf8") },
    });

    console.log(`[SecretManager] Created secret: ${secretId}`);
    return true;
  } catch (error: any) {
    if (error.message?.includes("already exists")) {
      console.log(`[SecretManager] Secret ${secretId} already exists`);
      return true;
    }
    console.error(`[SecretManager] Error creating ${secretId}:`, error);
    return false;
  }
}

// Convenience functions for common secrets
const secretCache: Map<string, string> = new Map();

async function getCachedSecret(
  secretId: string,
  envFallback: string
): Promise<string> {
  if (secretCache.has(secretId)) {
    return secretCache.get(secretId)!;
  }
  const value = await getSecret(secretId, envFallback);
  secretCache.set(secretId, value);
  return value;
}

export const getTwitchClientId = () =>
  getCachedSecret("twitch-client-id", "TWITCH_CLIENT_ID");

export const getTwitchClientSecret = () =>
  getCachedSecret("twitch-client-secret", "TWITCH_CLIENT_SECRET");

export const getDeepgramApiKey = () =>
  getCachedSecret("deepgram-api-key", "DEEPGRAM_API_KEY");

export const getClerkSecretKey = () =>
  getCachedSecret("clerk-secret-key", "CLERK_SECRET_KEY");

export function clearSecretCache(): void {
  secretCache.clear();
}

File: packages/shared/src/gcp/secret_manager.py
"""
Secret Manager Client - Secure credential retrieval for GCP services.

Provides a unified interface for accessing secrets with:
- Auto-detection of local development (uses .env)
- Caching to minimize API calls
- Version support for secret rotation
"""
import os
from typing import Optional
from functools import lru_cache


class SecretManagerClient:
    """
    Google Cloud Secret Manager client with local development fallback.
    
    In local development (when no GCP credentials are available),
    falls back to environment variables for secrets.
    """
    
    def __init__(self, project_id: Optional[str] = None):
        self.project_id = project_id or os.getenv("GCP_PROJECT_ID", "viralcue-local")
        self._client = None
        self._use_env_fallback = os.getenv("USE_SECRET_MANAGER", "false").lower() != "true"
    
    @property
    def client(self):
        """Lazy-load the Secret Manager client."""
        if self._client is None and not self._use_env_fallback:
            try:
                from google.cloud import secretmanager
                self._client = secretmanager.SecretManagerServiceClient()
            except Exception as e:
                print(f"[SecretManager] Failed to initialize, using env fallback: {e}")
                self._use_env_fallback = True
        return self._client
    
    def get_secret(
        self,
        secret_id: str,
        version: str = "latest",
        env_fallback: Optional[str] = None,
    ) -> str:
        """
        Retrieve a secret value.
        
        Args:
            secret_id: The secret ID in Secret Manager
            version: Version to retrieve (default: "latest")
            env_fallback: Environment variable name for local fallback
            
        Returns:
            The secret value as a string
        """
        # Local development fallback
        if self._use_env_fallback:
            env_var = env_fallback or secret_id.upper().replace("-", "_")
            value = os.getenv(env_var, "")
            if not value:
                print(f"[SecretManager] Warning: {env_var} not set in environment")
            return value
        
        # Build the secret version name
        name = f"projects/{self.project_id}/secrets/{secret_id}/versions/{version}"
        
        try:
            response = self.client.access_secret_version(request={"name": name})
            return response.payload.data.decode("UTF-8")
        except Exception as e:
            print(f"[SecretManager] Error accessing {secret_id}: {e}")
            # Fallback to environment variable
            env_var = env_fallback or secret_id.upper().replace("-", "_")
            return os.getenv(env_var, "")
    
    def create_secret(self, secret_id: str, secret_value: str) -> bool:
        """
        Create a new secret (for initialization/migration).
        
        Args:
            secret_id: The secret ID to create
            secret_value: The secret value
            
        Returns:
            True if created successfully
        """
        if self._use_env_fallback:
            print(f"[SecretManager] Skipping create in local mode: {secret_id}")
            return True
        
        parent = f"projects/{self.project_id}"
        
        try:
            # Create the secret
            self.client.create_secret(
                request={
                    "parent": parent,
                    "secret_id": secret_id,
                    "secret": {"replication": {"automatic": {}}},
                }
            )
            
            # Add the secret version
            secret_name = f"{parent}/secrets/{secret_id}"
            self.client.add_secret_version(
                request={
                    "parent": secret_name,
                    "payload": {"data": secret_value.encode("UTF-8")},
                }
            )
            
            print(f"[SecretManager] Created secret: {secret_id}")
            return True
            
        except Exception as e:
            if "already exists" in str(e).lower():
                print(f"[SecretManager] Secret {secret_id} already exists")
                return True
            print(f"[SecretManager] Error creating {secret_id}: {e}")
            return False


# Singleton instance
_client: Optional[SecretManagerClient] = None


def get_secret_manager(project_id: Optional[str] = None) -> SecretManagerClient:
    """Get the singleton Secret Manager client."""
    global _client
    if _client is None:
        _client = SecretManagerClient(project_id)
    return _client


# Convenience functions for common secrets
@lru_cache(maxsize=32)
def get_twitch_client_id() -> str:
    """Get Twitch Client ID."""
    return get_secret_manager().get_secret("twitch-client-id", env_fallback="TWITCH_CLIENT_ID")


@lru_cache(maxsize=32)
def get_twitch_client_secret() -> str:
    """Get Twitch Client Secret."""
    return get_secret_manager().get_secret("twitch-client-secret", env_fallback="TWITCH_CLIENT_SECRET")


@lru_cache(maxsize=32)
def get_deepgram_api_key() -> str:
    """Get Deepgram API Key."""
    return get_secret_manager().get_secret("deepgram-api-key", env_fallback="DEEPGRAM_API_KEY")


@lru_cache(maxsize=32)
def get_clerk_secret_key() -> str:
    """Get Clerk Secret Key."""
    return get_secret_manager().get_secret("clerk-secret-key", env_fallback="CLERK_SECRET_KEY")


def clear_secret_cache() -> None:
    """Clear the secret cache (use after rotation)."""
    get_twitch_client_id.cache_clear()
    get_twitch_client_secret.cache_clear()
    get_deepgram_api_key.cache_clear()
    get_clerk_secret_key.cache_clear()
